File: benchmarks/solutions/zig/r2-json-transform.zig
const std = @import("std");

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const stdout = std.io.getStdOut().writer();
    const stdin = std.io.getStdIn();

    const input = stdin.readToEndAlloc(allocator, 1024 * 1024) catch |e| {
        std.debug.print("error: {}\n", .{e});
        std.process.exit(1);
    };

    const parsed = std.json.parseFromSlice(std.json.Value, allocator, input, .{}) catch {
        std.debug.print("error: invalid JSON\n", .{});
        std.process.exit(1);
    };

    const users = parsed.value.array;

    try stdout.writeAll("[");
    var first = true;
    for (users.items) |item| {
        const obj = item.object;
        const active = obj.get("active") orelse continue;
        const age = obj.get("age") orelse continue;
        if (active != .true or age.integer <= 18) continue;

        const name = obj.get("name") orelse continue;
        const email = obj.get("email") orelse continue;

        if (!first) try stdout.writeAll(",");
        try stdout.print("{{\"name\":\"{s}\",\"email\":\"{s}\"}}", .{
            name.string, email.string,
        });
        first = false;
    }
    try stdout.writeAll("]\n");
}

File: benchmarks/solutions/c/r2-json-transform.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <cjson/cJSON.h>

int main(void) {
    char buf[65536];
    size_t len = fread(buf, 1, sizeof(buf) - 1, stdin);
    buf[len] = '\0';

    cJSON *users = cJSON_Parse(buf);
    if (!users) {
        fprintf(stderr, "error: invalid JSON\n");
        return 1;
    }

    cJSON *result = cJSON_CreateArray();
    cJSON *u;
    cJSON_ArrayForEach(u, users) {
        cJSON *active = cJSON_GetObjectItem(u, "active");
        cJSON *age = cJSON_GetObjectItem(u, "age");
        if (cJSON_IsTrue(active) && age->valueint > 18) {
            cJSON *out = cJSON_CreateObject();
            cJSON_AddStringToObject(out, "name",
                cJSON_GetObjectItem(u, "name")->valuestring);
            cJSON_AddStringToObject(out, "email",
                cJSON_GetObjectItem(u, "email")->valuestring);
            cJSON_AddItemToArray(result, out);
        }
    }

    char *json = cJSON_PrintUnformatted(result);
    puts(json);
    free(json);
    cJSON_Delete(users);
    cJSON_Delete(result);
    return 0;
}
